### PR TITLE
HDDS-8436. Support setSafeMode(), isFileClosed() FileSystem API

### PR DESCRIPTION
**Pull request by jojochuang**
## What changes were proposed in this pull request?

Implement new Hadoop file system APIs to support HBase.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8436 Support setSafeMode(), isFileClosed() FileSystem API

## How was this patch tested?
Unit tests

### CODE REVIEW

File: hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto [hunk 1]
```diff
@@ -135,6 +135,14 @@ enum Type {
   SnapshotPurge = 118;
   RecoverLease = 119;
   SetTimes = 120;
+  SetSafeMode = 121;
```
Comment (szetszwo) [inline hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:138]: @jojochuang , 121 is already taken in the master branch.  Please sync this with the master branch.
```
<<<<<<< HEAD
  RefetchSecretKey = 121;
=======
  SetSafeMode = 121;
>>>>>>> jojochuang/HDDS-8436
```

**Review by szetszwo — COMMENTED** [review]
@jojochuang , thanks a lot for working on this!  The change looks good.  Please see the comments inlined.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java [hunk 2]
```diff
@@ -4431,6 +4433,24 @@ public void setTimes(OmKeyArgs keyArgs, long mtime, long atime)
...
 
+  @Override
+  public boolean setSafeMode(SafeModeAction action, boolean isChecked)
+      throws IOException {
+    switch (action) {
+    case ENTER:
+      throw new OMException("Enter safe mode is unsupported",
+          INTERNAL_ERROR);
+    case LEAVE:
+    case FORCE_EXIT:
```
[the remark this answers is not in the document]
Comment (szetszwo) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:4444]: For `LEAVE`, it should exit safe mode only if preCheck is true.  We should make the change separately.
```java
//SCMSafeModeManager
@@ -237,20 +237,25 @@ public void completePreCheck(EventPublisher eventQueue) {
    * 2. Emits START_REPLICATION for ReplicationManager.
    * 3. Cleanup resources.
    * 4. Emit safe mode status.
-   * @param eventQueue
    */
   @VisibleForTesting
-  public void exitSafeMode(EventPublisher eventQueue) {
-    LOG.info("SCM exiting safe mode.");
-    // If safemode is exiting, then pre check must also have passed so
-    // set it to true.
-    setPreCheckComplete(true);
+  public boolean exitSafeMode(boolean force) {
+    LOG.info("SCM exiting safe mode, force? {}", force);
+    if (force) {
+      // If force exiting, set preCheck to true.
+      setPreCheckComplete(true);
+    } else {
+      if (!getPreCheckComplete()) {
+        return false;
+      }
+    }
     setInSafeMode(false);
 
     // TODO: Remove handler registration as there is no need to listen to
     // register events anymore.
 
     emitSafeModeStatus();
+    return true;
   }
```

File: hadoop-ozone/ozonefs-hadoop3/src/main/java/org/apache/hadoop/fs/ozone/RootedOzoneFileSystem.java [hunk 3]
```diff
@@ -118,4 +121,39 @@ public boolean hasPathCapability(final Path path, final String capability)
...
+    LOG.trace("isFileClosed() path:{}", f);
+    Path qualifiedPath = makeQualified(f);
+    String key = pathToKey(qualifiedPath);
+    return getAdapter().isFileClosed(key);
+  }
+
+  @Override
+  public boolean setSafeMode(SafeModeAction action, boolean isChecked)
+      throws IOException {
+    statistics.incrementWriteOps(1);
```
Comment (szetszwo) [inline hadoop-ozone/ozonefs-hadoop3/src/main/java/org/apache/hadoop/fs/ozone/RootedOzoneFileSystem.java:155]: Check if the action is not `GET` before `incrementWriteOps`.

**Review by szetszwo — COMMENTED** [review]
@jojochuang , thanks for the update!

We need to check `GET` before `incrementWriteOp` in  two of `RootedOzoneFileSystem`s.  BTW, is it possible to move the `setSafeMode` method to `Basic RootedOzoneFileSystem`?

**Comment by jojochuang** [comment]
> BTW, is it possible to move the setSafeMode method to Basic RootedOzoneFileSystem?

We can't, because Hadoop 2 does not have SafeMode interface (introduced in Hadoop 3.3.6). So I can only put it where Ozone is not run with Hadoop 2 runtime.

**Comment by jojochuang** [comment]
But as it turns out, we test MR jobs with Hadoop 3.1 3.2 and 3.3 runtime, and the MR job tests fail with Hadoop 3.1 and 3.2 (due to the lacking of SafeMode and LeaseRecoverable interface). So either we find a workaround or remove the Hadoop 3.1 3.2 tests.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**